doc/source/dev/gitwash/development_workflow.rst

```diff
@@ -195,27 +187,78 @@ Standard acronyms to start the commit message with are::
    REL: related to releasing numpy
 
 
+.. _asking-for-merging:
+
+Asking for your changes to be merged with the main repo
+=======================================================
+
+When you feel your work is finished, you can file a pull request and
+submit an optional code review. Code reviews are typically submitted when an
+API change is proposed, such as adding a new function or modifying an option
+for an existing function.
+
+Github has a help page that outlines the process for `filing pull requests`_.
+If you are going to file a code review, copy the URL for the `pull request
+review`_ between your branch and ``master``, and post it to the `NumPy mailing
+list`_ asking for review. 
+
+.. note:: 
+
+    Make sure your comparison is with ``master`` at the point where you
+    created your working branch, which is indicated by an ellipsis (``...``).
+    The importance of this notation is outlined in :ref:`dot2-dot3`.
+
+
+.. _pushing-to-main:
+
+Pushing changes to the main repo
+================================
+
+When you have a set of "ready" changes in a feature branch ready for
+Numpy's ``master`` or ``maintenance`` branches, you can push
+them to ``upstream`` as follows:
+
+1. First, merge or rebase on the target branch.
+
+   a) Only a few commits: prefer rebasing::
+
+        git fetch upstream
+        git rebase upstream/master
+
+      See :ref:`rebasing-on-master`.
+
+   b) Many related commits: consider creating a merge commit::
+
+        git fetch upstream
+        git merge --no-ff upstream/master
+
+2. Check that what you are going to push looks sensible::
+
+        git log -p upstream/master..
+        git log --oneline --graph
+
+3. Push to upstream::
+
+        git push upstream my-feature-branch:master
```

Comment:
That's right, we've pretty much settled on always making merge commits. When we started on github we tried to keep the history as one linear line, but eventually figured out that that was an abuse of git and that merges made it easier to track and deal with multiple commits, and even for single commits it allows determining who did the merge, as opposed to who made the modifications.